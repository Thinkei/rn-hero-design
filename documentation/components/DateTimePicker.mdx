---
name: DateTimePicker
menu: Components
---

import { Props } from 'gatsby-theme-docz/src/components/Props';
import { DateTimePicker } from 'rn-hero-design'

# DateTimePicker

## Properties

<Props props_={{
  show: { type: 'Boolean', required: false, defaultValue: { value: 'false' } },
  mode: { type: '"date" | "time"', required: false, defaultValue: { value: '"date"' } },
  value: { type: 'DateObject', required: false, defaultValue: { value: 'new Date()' } },
  onChange: { type: 'DateObject => void', required: false },
  onDismiss: { type: 'void => void', required: false },
  theme: { type: 'Object', required: false, description: 'The theme object contains style of all components. Can be injected by `injectTheme` HOC' },
}} />

