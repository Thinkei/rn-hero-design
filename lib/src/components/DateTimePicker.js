import React from 'react';
import t from 'prop-types';
import {
  View,
  Text,
  TouchableOpacity,
  DatePickerAndroid,
  DatePickerIOS,
  Keyboard,
  StyleSheet,
} from 'react-native';
import SafeAreaView from 'react-native-safe-area-view';
import heroTheme from '../themes/hero';
import { isAndroid } from '../helpers/platform';

class DateTimePickerIOS extends React.Component {
  constructor(props) {
    super(props);
    this.state = { pickedDate: props.value };
  }

  componentDidMount() {
    if (this.props.show) Keyboard.dismiss();
    this.keyboardWillShowListener = Keyboard.addListener(
      'keyboardWillShow',
      this.handleKeyboardWillShow,
    );
  }

  componentWillUnmount() {
    this.keyboardWillShowListener.remove();
  }

  componentWillReceiveProps({ value, show }) {
    const prevShow = this.props.show;
    if (prevShow !== show && show) Keyboard.dismiss();
    this.setState({ pickedDate: value });
  }

  handleKeyboardWillShow = () => this.props.onDismiss();

  handleDismiss = () => this.props.onDismiss();

  handleChange = () => {
    const { pickedDate } = this.state;
    const { onChange, onDismiss } = this.props;
    onChange(pickedDate);
    onDismiss();
  };

  handleDateChange = date => this.setState({ pickedDate: date });

  render() {
    const { pickedDate } = this.state;
    const { show, mode, theme = heroTheme } = this.props;

    if (!show) return null;

    return (
      <SafeAreaView
        forceInset={{ bottom: 'always' }}
        style={theme.dateTimePicker.wrapper}
      >
        <View style={theme.dateTimePicker.actions}>
          <TouchableOpacity
            onPress={this.handleDismiss}
            style={theme.dateTimePicker.action}
          >
            <Text style={theme.dateTimePicker.actionText}>Cancel</Text>
          </TouchableOpacity>
          <TouchableOpacity
            onPress={this.handleChange}
            style={theme.dateTimePicker.action}
          >
            <Text
              style={StyleSheet.flatten([
                theme.dateTimePicker.actionText,
                { fontWeight: '600' },
              ])}
            >
              Done
            </Text>
          </TouchableOpacity>
        </View>
        <DatePickerIOS
          mode={mode}
          date={pickedDate}
          onDateChange={this.handleDateChange}
        />
      </SafeAreaView>
    );
  }
}

class DateTimePickerAndroid extends React.Component {
  componentDidMount() {
    const { show, mode, value, onChange, onDismiss } = this.props;

    if (show)
      this.openDateTimePickerAndroid({ mode, value, onChange, onDismiss });
  }

  componentWillReceiveProps(nextProps) {
    const prevShow = this.props.show;
    const { show, mode, value, onChange, onDismiss } = nextProps;

    if (prevShow !== show && show)
      this.openDateTimePickerAndroid({ mode, value, onChange, onDismiss });
  }

  openDateTimePickerAndroid = ({ mode, value, onChange, onDismiss }) => {
    if (mode === 'date') {
      DatePickerAndroid.open({
        date: value,
      }).then(({ action, year, month, day }) => {
        if (action === DatePickerAndroid.dateSetAction)
          onChange(new Date(year, month, day));

        onDismiss();
      });
    }
  };

  render() {
    return null;
  }
}

const DateTimePicker = props =>
  isAndroid ? (
    <DateTimePickerAndroid {...props} />
  ) : (
    <DateTimePickerIOS {...props} />
  );

DateTimePicker.propTypes = {
  show: t.bool,
  mode: t.oneOf(['date', 'datetime']),
  value: t.object,
  onChange: t.func,
  onDismiss: t.func,
};

DateTimePicker.defaultProps = {
  mode: 'date',
};

export default DateTimePicker;
