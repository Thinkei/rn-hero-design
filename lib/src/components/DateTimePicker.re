open ReactNative;

let noop = _ => ();

module DateTimePickerIOS = {
  [@react.component]
  let make = (~show, ~mode, ~value, ~onChange, ~onDismiss, ~theme) => {
    let (pickedDate, setPickedDate) = React.useState(() => value);

    let handleSelect =
      React.useCallback1(
        _ => {
          onChange(pickedDate);
          onDismiss();
        },
        [|pickedDate|],
      );

    let handleCancel = React.useCallback(_ => onDismiss());

    let handleKeyboardWillShow = React.useCallback(_ => onDismiss());

    let handleChange =
      (. _, date) =>
        switch (Js.Nullable.toOption(date)) {
        | Some(date) => setPickedDate(_ => date)
        | None => ()
        };

    React.useEffect2(
      _ => {
        if (show) {
          Keyboard.dismiss();
        };
        let keyboardWillShowListener =
          Keyboard.addListener(`keyboardWillShow, handleKeyboardWillShow);
        setPickedDate(_ => value);
        Some(() => EventSubscription.remove(keyboardWillShowListener));
      },
      (show, value),
    );

    if (!show) {
      React.null;
    } else {
      <RNSafeAreaView
        forceInset={"bottom": "always", "top": "never"}
        style=theme##dateTimePicker##wrapper>
        <View style=theme##dateTimePicker##actions>
          <TouchableOpacity
            onPress=handleCancel style=theme##dateTimePicker##action>
            <Text style=theme##dateTimePicker##actionText>
              "Cancel"->React.string
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            onPress=handleSelect style=theme##dateTimePicker##action>
            <Text
              style={StyleSheet.flatten([|
                theme##dateTimePicker##actionText,
                Style.style(~fontWeight=`_600, ()),
              |])}>
              "Done"->React.string
            </Text>
          </TouchableOpacity>
        </View>
        <RNCDateTimePicker
          mode
          display=`spinner
          value=pickedDate
          onChange=handleChange
        />
      </RNSafeAreaView>;
    };
  };
};

module DateTimePickerAndroid = {
  [@react.component]
  let make = (~show, ~mode, ~value, ~onChange, ~onDismiss) => {
    let handleChange =
      (. _, date) => {
        switch (Js.Nullable.toOption(date)) {
        | Some(date) =>
          onDismiss();
          onChange(date);
        | None => onDismiss()
        };
      };

    if (!show) {
      React.null;
    } else {
      <RNCDateTimePicker mode display=`default value onChange=handleChange />;
    };
  };
};

[@react.component]
let make =
    (
      ~show=false,
      ~value: Js.Date.t=Js.Date.make(),
      ~mode=`date,
      ~onChange=noop,
      ~onDismiss=noop,
      ~theme=Hero_Theme.default,
    ) =>
  Helpers.Platform.isAndroid
    ? <DateTimePickerAndroid show mode value onChange onDismiss />
    : <DateTimePickerIOS show mode value onChange onDismiss theme />;

let default = Helpers.injectTheme(make);
