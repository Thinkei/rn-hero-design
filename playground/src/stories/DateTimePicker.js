import React, { useState } from 'react';
import { View, TouchableOpacity } from 'react-native';
import {
  DateTimePicker,
  TextInput,
  Container,
  ReDateTimePicker,
} from 'rn-hero-design';

const noop = () => {};

const DateTimePickerScreen = () => {
  const today = new Date();
  const [startDate, setStartDate] = useState(today);
  const [endDate, setEndDate] = useState(today);
  const [changingField, setChangingField] = useState(null);

  const fieldMap = {
    startDate: {
      value: startDate,
      onChange: setStartDate,
    },
    endDate: {
      value: endDate,
      onChange: setEndDate,
    },
    null: {
      value: today,
      onChange: noop,
    },
  };

  return (
    <>
      <Container fluid>
        <TouchableOpacity onPress={() => setChangingField('startDate')}>
          <View pointerEvents="none">
            <TextInput
              label="Start date"
              value={startDate.toISOString()}
              onChangeText={noop}
            />
          </View>
        </TouchableOpacity>

        <TouchableOpacity onPress={() => setChangingField('endDate')}>
          <View pointerEvents="none">
            <TextInput
              label="End date"
              value={endDate.toISOString()}
              onChangeText={noop}
            />
          </View>
        </TouchableOpacity>
      </Container>

      <ReDateTimePicker
        show={!!changingField}
        value={fieldMap[changingField].value}
        onChange={date => fieldMap[changingField].onChange(date)}
        onDismiss={() => setChangingField(null)}
      />
    </>
  );
};

export default DateTimePickerScreen;
