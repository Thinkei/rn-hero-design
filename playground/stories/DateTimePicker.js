import React, { useState } from 'react';
import { View, TouchableOpacity } from 'react-native';
import { DateTimePicker, TextInput, Container } from 'rn-hero-design';

const noop = () => {};

const DateTimePickerScreen = () => {
  const [startDate, setStartDate] = useState(new Date());
  const [endDate, setEndDate] = useState(new Date());
  const [changingField, setChangingField] = useState(null);

  const fieldMap = {
    startDate: {
      value: startDate,
      onChange: setStartDate,
    },
    endDate: {
      value: endDate,
      onChange: setEndDate,
    },
    null: {
      value: null,
      onChange: noop,
    },
  };

  return (
    <>
      <Container fluid>
        <TouchableOpacity onPress={() => setChangingField('startDate')}>
          <View pointerEvents="none">
            <TextInput
              label="Start date"
              value={startDate.toISOString()}
              onChangeText={noop}
            />
          </View>
        </TouchableOpacity>

        <TouchableOpacity onPress={() => setChangingField('endDate')}>
          <View pointerEvents="none">
            <TextInput
              label="End date"
              value={endDate.toISOString()}
              onChangeText={noop}
            />
          </View>
        </TouchableOpacity>
      </Container>

      <DateTimePicker
        show={!!changingField}
        value={fieldMap[changingField].value}
        onChange={date => fieldMap[changingField].onChange(date)}
        onDismiss={() => setChangingField(null)}
      />
    </>
  );
};

export default DateTimePickerScreen;
