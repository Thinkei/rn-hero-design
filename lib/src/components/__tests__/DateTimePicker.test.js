import React from 'react';
import DateTimePicker from '../DateTimePicker.gen';
import renderer from 'react-test-renderer';
import * as ReactNative from 'react-native';

jest.mock('react-native/Libraries/Components/Keyboard/Keyboard', () => ({
  addListener: jest.fn(),
  dismiss: jest.fn(),
}));
describe('DateTimePicker', () => {
  it.each`
    element
    ${<DateTimePicker show={true} />}
    ${<DateTimePicker show={false} />}
    ${<DateTimePicker show={true} mode="date" />}
    ${<DateTimePicker show={true} mode="time" />}
  `('renders correctly', ({ element }) => {
    const tree = renderer.create(element).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
