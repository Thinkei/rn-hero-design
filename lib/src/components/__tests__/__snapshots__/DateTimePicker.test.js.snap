// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker renders correctly 1`] = `
<View
  forwardedRef={[Function]}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#f5f6f8",
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#d9dbdf",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
          }
        }
      >
        Cancel
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
            "fontWeight": "600",
          }
        }
      >
        Done
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`DateTimePicker renders correctly 2`] = `null`;

exports[`DateTimePicker renders correctly 3`] = `
<View
  forwardedRef={[Function]}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#f5f6f8",
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#d9dbdf",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
          }
        }
      >
        Cancel
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
            "fontWeight": "600",
          }
        }
      >
        Done
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`DateTimePicker renders correctly 4`] = `
<View
  forwardedRef={[Function]}
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#f5f6f8",
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#d9dbdf",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
          }
        }
      >
        Cancel
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      onPress={[Function]}
      style={
        Object {
          "height": 45.9875,
          "justifyContent": "center",
          "paddingHorizontal": 16.25,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1dbeee",
            "fontSize": 16.25,
            "fontWeight": "600",
          }
        }
      >
        Done
      </Text>
    </TouchableOpacity>
  </View>
  <RNDateTimePicker
    date={1628467200000}
    displayIOS="spinner"
    mode="time"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 216,
      }
    }
  />
</View>
`;
